pkg/node/node.go

```diff
@@ -173,3 +178,91 @@ func IsMachineReady(machine *v3.Node) bool {
 	}
 	return false
 }
+
+func DrainBeforeDelete(node *v3.Node, cluster *v3.Cluster) bool {
+	// never drain an etcd node, you will regret it
+	if IsEtcd(node.Status.NodeConfig) {
+		return false
+	}
+
+	// vsphere nodes with cloud providers need to be drained to unmount vmdk files as vsphere deletes them automatically
+	if nil != cluster.Spec.RancherKubernetesEngineConfig.CloudProvider.VsphereCloudProvider {
```

Comment:
Is it possible for CloudProvider to ever be nil here?

Reply:
ya it's not a pointer, that's why it worked and VsphereCloudProvider is a pointer hence the nil check.